internal/apis/certmanager/types_certificate.go

```diff
@@ -531,3 +537,41 @@ type CertificateSecretTemplate struct {
 	// +optional
 	Labels map[string]string
 }
+
+// NameConstraints is a type to represent x509 NameConstraints
+type NameConstraints struct {
+	// Permitted contains the constraints in which the names must be located.
+	//
+	// +optional
+	Permitted *NameConstraintItem
+	// Excluded contains the constraints which must be disallowed. Any name matching a
+	// restriction in the excluded field is invalid regardless
+	// of information appearing in the permitted
+	//
+	// +optional
+	Excluded *NameConstraintItem
+}
+
+type NameConstraintItem struct {
+	// DNSDomains is a list of DNS domains that are permitted or excluded.
+	//
+	// +optional
+	DNSDomains []string
+	// IPRanges is a list of IP Ranges that are permitted or excluded.
+	// This should be a valid CIDR notation.
+	//
+	// +optional
+	IPRanges []string
+	// EmailAddresses is a list of Email Addresses that are permitted or excluded.
+	//
+	// +optional
+	EmailAddresses []string
+	// URIDomains is a list of URI domains that are permitted or excluded.
+	//
+	// +optional
+	URIDomains []string
+	// if true then the name constraints are marked critical.
+	//
+	// +optional
+	Critical bool
```

Comment:
suggestion: I don't think this should be under NameConstraintItem, I think it should be under NameConstraints

The API should look like this IMO:

```yaml
nameConstraints:
    critical: true
    permitted:
      dnsDomains: ["example1.com", "example2.com"]
      ipRanges: ["10.10.0.0/16"]
      emailAddress: ["example@example.org"]
    excluded:
      ipRanges: ["10.10.0.0/24"]
```

Reply:
Currently critical is supported fir permittedDNSDomains only in go x509 lib. I had it in the NameConstraintItem if in future it supports different knob for excluded. I can move it to `nameConstraints` if it does not make sense though.